public/services/data_source_service.ts

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { BehaviorSubject, PartialObserver, Subscription } from 'rxjs';
+import type { IUiSettingsClient } from '../../../../src/core/public';
+import type { DataSourceOption } from '../../../../src/plugins/data_source_management/public/components/data_source_menu/types';
+import type { DataSourceManagementPluginSetup } from '../types';
+
+export enum DataSourceIdFrom {
+  UiSettings,
+  DataSourceSelection,
+  Customized,
+}
+
+const getSingleSelectedDataSourceOption = (
+  dataSourceSelection: Map<string, DataSourceOption[]>
+) => {
+  const values = [...dataSourceSelection.values()];
+  // Should use default index if multi data source selected
+  if (values.length === 0 || values.length > 1 || values?.[0]?.length > 1) {
+    return null;
+  }
+  return values[0][0];
+};
+
+export class DataSourceService {
+  private dataSourceId$ = new BehaviorSubject<string | null>(null);
+  private uiSettings: IUiSettingsClient | undefined;
+  private dataSourceManagement: DataSourceManagementPluginSetup | undefined;
+  private dataSourceSelectionSubscription: Subscription | undefined;
+  private dataSourceIdFrom: DataSourceIdFrom | undefined;
+
+  constructor() {}
+
+  initDefaultDataSourceIdIfNeed() {
```

Comment:
When will this function to be called?

Reply:
> so chat bot can init data source id when dialog open.

Found in the PR description, but do chatbot really need to call `initDefaultDataSourceIdIfNeed` every time it's opened? I feel it more intuitive if simply `subscribeDataSourceId` or `getDataSourceId` can do it.

Reply:
The motivation of init everytime when chatbot open is get latest default data source id. Since the default data source id may changed in another tabs, add the init code to chatbot open to avoid using the expired default data source id.